File: src/service/account_service.go
/*
 * Auth: account_service.go
 * Copyright (C) 2025 mtctx
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the **GNU General Public License** as published
 * by the Free Software Foundation, either **version 3** of the License, or
 * (at your option) any later version.
 *
 * *This program is distributed WITHOUT ANY WARRANTY;** see the
 * GNU General Public License for more details, which you should have
 * received with this program.
 *
 * SPDX-FileCopyrightText: 2025 mtctx
 * SPDX-License-Identifier: GPL-3.0-only
 */

package service

import (
	"crypto/rand"
	"encoding/base32"
	"errors"
	"src/crypto"
	"src/data"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
)

func GetAccount(setup GeneralData[data.Account], uuid uuid.UUID) (data.Account, error) {
	err := setup.Validate()
	if err != nil {
		return data.Account{}, err
	}

	if value, found := setup.Cache.Get(uuid[:]); found {
		return value, nil
	}

	const query = "SELECT * FROM accounts WHERE uuid = $1"

	var account data.Account
	err = setup.Database.QueryRow(setup.DBContext, query, uuid).Scan(&account)
	if err != nil {
		return data.Account{}, err
	}

	return account, nil
}

func CheckPassword(setup GeneralData[data.Account], uuid uuid.UUID, unhashedPassword string) (bool, error) {
	err := setup.Validate()
	if err != nil {
		return false, err
	}

	const query = "SELECT password_hash FROM accounts WHERE uuid = $1"
	var passwordHash string
	err = setup.Database.QueryRow(setup.DBContext, query, uuid).Scan(&passwordHash)
	if err != nil {
		return false, err
	}

	if strings.TrimSpace(passwordHash) == "" {
		return false, nil
	}

	result, err := crypto.VerifyArgon2id(unhashedPassword, passwordHash)
	if err != nil {
		return false, err
	}

	return result, nil
}

func UpdatePassword(setup GeneralData[data.Account], uuid uuid.UUID, newPassword string) error {
	err := setup.Validate()
	if err != nil {
		return err
	}

	hashedPassword, err := crypto.HashArgon2id(newPassword)
	if err != nil {
		return err
	}
	query := "UPDATE accounts SET password_hash = $1 WHERE uuid = $2; DELETE FROM sessions WHERE user_uuid = $2"
	_, err = setup.Database.Exec(setup.DBContext, query, hashedPassword, uuid)
	return err
}

func AccountLogin(setup GeneralData[data.Account], sss GeneralData[data.Session], validFor time.Duration, username string, password string) (string, error) {
	err := setup.Validate()
	if err != nil {
		return "", err
	}

	const query = "SELECT uuid, password_hash FROM accounts WHERE username = $1"
	var (
		queriedUUID    uuid.UUID
		hashedPassword string
	)
	err = setup.Database.QueryRow(setup.DBContext, query, username).Scan(&queriedUUID, &hashedPassword)
	if err != nil {
		return "", err
	}

	result, err := crypto.VerifyArgon2id(password, hashedPassword)
	if err != nil {
		return "", err
	}

	if !result {
		return "", errors.New("invalid credentials")
	}

	session, mac, err := CreateSession(sss, queriedUUID, validFor)
	if err != nil {
		return "", err
	}

	return BuildSessionString(session.UUID, mac), nil
}

func CreateAccount(setup GeneralData[data.Account], username string, password string) (data.Account, error) {
	err := setup.Validate()
	if err != nil {
		return data.Account{}, err
	}

	generatedUUID := uuid.New()
	hashedPassword, err := crypto.HashArgon2id(password)
	if err != nil {
		return data.Account{}, err
	}
	webhookID := uuid.New()
	recoveryCodes, err := GenerateAccountRecoveryCodes(5)
	if err != nil {
		return data.Account{}, err
	}

	query := "INSERT INTO accounts (uuid, username, password_hash, webhook_id, recovery_codes) VALUES ($1, $2, $3, $4, $5)"
	_, err = setup.Database.Exec(setup.DBContext, query, generatedUUID, username, hashedPassword, webhookID, recoveryCodes)
	if err != nil {
		return data.Account{}, err
	}

	account := data.Account{
		UUID:          generatedUUID,
		Username:      username,
		PasswordHash:  hashedPassword,
		WebhookID:     webhookID,
		RecoveryCodes: recoveryCodes,
	}
	added := setup.Cache.Set(generatedUUID[:], account, 0)
	if !added {
		log.Error().Msg("Failed to add session to cache")
	}

	return account, nil
}

func GenerateAccountRecoveryCodes(n int) ([]string, error) {
	codes := make([]string, n)
	for i := range codes {
		b := make([]byte, 10)
		if _, err := rand.Read(b); err != nil {
			return nil, err
		}
		codes[i] = strings.ToUpper(base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(b))
		// e.g. X7K9P2M4Q1R8
	}
	return codes, nil
}

func DeleteAccount(setup GeneralData[data.Account], uuid uuid.UUID) error {
	err := setup.Validate()
	if err != nil {
		return err
	}

	setup.Cache.Del(uuid[:])

	query := "DELETE FROM accounts WHERE uuid = $1"
	_, err = setup.Database.Exec(setup.DBContext, query, uuid)
	return err
}

File: src/main/kotlin/dev/vayen/service/AccountService.kt
/*
 * Vayen Auth (Vayen_Auth.main): AccountService.kt
 * Copyright (C) 2025 mtctx
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the **GNU General Public License** as published
 * by the Free Software Foundation, either **version 3** of the License, or
 * (at your option) any later version.
 *
 * *This program is distributed WITHOUT ANY WARRANTY;** see the
 * GNU General Public License for more details, which you should have
 * received with this program.
 *
 * SPDX-FileCopyrightText: 2025 mtctx
 * SPDX-License-Identifier: GPL-3.0-only
 */

package dev.vayen.service

import dev.vayen.data.Account
import dev.vayen.other.GeneralData
import dev.vayen.table.AccountTable
import dev.vayen.table.SessionTable
import dev.vayen.table.mfa.MFATOTPTable
import dev.vayen.table.mfa.MFAWebAuthnTable
import mtctx.utilities.*
import mtctx.utilities.crypto.Argon2
import mtctx.utilities.crypto.SECURE_RANDOM
import org.bouncycastle.util.encoders.Base32
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import java.util.*

class AccountService(private val setup: GeneralData<UUID, Account>) : DatabaseService<UUID, Account>() {
    init {
        SchemaUtils.create(AccountTable, MFATOTPTable, MFAWebAuthnTable)
    }

    private fun <T> get(column: ExpressionWithColumnType<T>, type: T): Account? =
        AccountTable.selectAll().where { column eq type }.map {
            Account(
                uuid = it[AccountTable.uuid],
                username = it[AccountTable.username],
                passwordHash = it[AccountTable.passwordHash],
                webhookUUID = it[AccountTable.webhookUUID],
                recoveryCodes = it[AccountTable.recoveryCodes]
            )
        }.singleOrNull()

    override suspend operator fun get(
        id: UUID
    ): Outcome<Account> = runCatchingOutcomeOf<Account> {
        val account =
            setup.cache.get(id)
                ?: dbQuery { get(AccountTable.uuid, id) }
                ?: return failure("Account not found")
        account
    }

    suspend fun getByUsername(username: String): Outcome<Account> =
        dbQuery { get(AccountTable.username, username) }?.let { success(it) } ?: failure("Account not found")

    suspend fun getByWebhookUUID(webhookUUID: UUID): Outcome<Account> =
        dbQuery { get(AccountTable.webhookUUID, webhookUUID) }?.let { success(it) } ?: failure("Account not found")

    suspend fun getBySession(sessionUUID: UUID): Outcome<Account> = runCatchingOutcomeOf<Account> {
        dbQuery {
            (SessionTable innerJoin AccountTable)
                .selectAll()
                .where { SessionTable.uuid eq sessionUUID }
                .map {
                    Account(
                        uuid = it[AccountTable.uuid],
                        username = it[AccountTable.username],
                        passwordHash = it[AccountTable.passwordHash],
                        webhookUUID = it[AccountTable.webhookUUID],
                        recoveryCodes = it[AccountTable.recoveryCodes]
                    )
                }.singleOrNull()
        } ?: return failure("No user found for the given session UUID")
    }

    suspend fun recoveryCodes(uuid: UUID): Outcome<List<String>> = this[uuid].mapCatching { it.recoveryCodes }

    suspend fun checkPassword(uuid: UUID, unhashedPassword: String): Outcome<Boolean> =
        this[uuid].mapCatching { return Argon2.verify(unhashedPassword, it.passwordHash) }

    suspend fun updatePassword(uuid: UUID, newPassword: String): Outcome<Unit> = runCatchingOutcomeOf<Unit> {
        dbQuery {
            AccountTable.update({ AccountTable.uuid eq uuid }) {
                it[AccountTable.passwordHash] = Argon2.hash(newPassword).hash
            }
            SessionTable.deleteWhere { SessionTable.userUUID eq uuid }
        }
    }

    suspend fun login(sessionService: SessionService, username: String, password: String): Outcome<String> =
        getByUsername(username).mapCatching { account ->
            val correctPassword = Argon2.verify(password, account.passwordHash).getOrNull()
                ?: return failure("Could not check passwords!")
            if (!correctPassword) return failure("Incorrect password!")

            return sessionService.create(account.uuid)
        }

    suspend fun create(username: String, password: String): Outcome<Unit> = runCatchingOutcomeOf<Unit> {
        dbQuery {
            if (AccountTable.selectAll().where { AccountTable.username eq username }.limit(1)
                    .any()
            ) throw IllegalStateException("Username is taken.")

            AccountTable.insert {
                it[AccountTable.uuid] = UUID.randomUUID()
                it[AccountTable.username] = username
                it[AccountTable.passwordHash] = Argon2.hash(password).hash
                it[AccountTable.webhookUUID] = UUID.randomUUID()
                it[AccountTable.recoveryCodes] = generateAccountRecoveryCodes(5)
            }
        }
    }


    override suspend fun delete(id: UUID): Outcome<Unit> = runCatchingOutcomeOf<Unit> {
        setup.cache.remove(id)
        dbQuery {
            AccountTable.deleteWhere { AccountTable.uuid eq id }
            SessionTable.deleteWhere { SessionTable.userUUID eq id }
        }
    }

    companion object {
        fun generateAccountRecoveryCodes(n: Int): List<String> {
            return List(n) {
                val bytes = ByteArray(10)
                SECURE_RANDOM.nextBytes(bytes)
                Base32.toBase32String(bytes).uppercase()
            }
        }
    }
}
